Mathlib/Combinatorics/Quiver/Symmetric.lean

```diff
@@ -0,0 +1,247 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn, Antoine Labelle, Rémi Bottinelli
+Ported by: Joël Riou, Rémi Bottinelli
+-/
+import Mathlib.Combinatorics.Quiver.Subquiver
+import Mathlib.Combinatorics.Quiver.Path
+import Mathlib.Combinatorics.Quiver.Push
+import Mathlib.Data.Sum.Basic
+
+/-!
+## Symmetric quivers and arrow reversal
+
+This file contains constructions related to symmetric quivers:
+
+* `Symmetrify V` adds formal inverses to each arrow of `V`.
+* `HasReverse` is the class of quivers where each arrow has an assigned formal inverse.
+* `HasInvolutiveReverse` extends `HasReverse` by requiring that the reverse of the reverse
+  is equal to the original arrow.
+* `Prefunctor.PreserveReverse` is the class of prefunctors mapping reverses to reverses.
+* `Symmetrify.of`, `Symmetrify.lift`, and the associated lemmas witness the universal property
+  of `Symmetrify`.
+-/
+
+universe v u
+
+namespace Quiver
+
+/-- A type synonym for the symmetrized quiver (with an arrow both ways for each original arrow).
+    NB: this does not work for `Prop`-valued quivers. It requires `[Quiver.{v+1} V]`. -/
+-- Porting note: no hasNonemptyInstance linter yet
+def Symmetrify (V : Type _) := V
+
+instance symmetrifyQuiver (V : Type _) [Quiver V] : Quiver (Symmetrify V) :=
+  ⟨fun a b : V ↦ Sum (a ⟶ b) (b ⟶ a)⟩
+
+variable (U V W : Type) [Quiver.{u + 1} U] [Quiver.{v + 1} V]
+
+/-- A quiver `HasReverse` if we can reverse an arrow `p` from `a` to `b` to get an arrow
+    `p.reverse` from `b` to `a`.-/
+class HasReverse where
+  /-- the map which sends an arrow to its reverse -/
+  reverse' : ∀ {a b : V}, (a ⟶ b) → (b ⟶ a)
+
+/-- Reverse the direction of an arrow. -/
+def reverse {V} [Quiver.{v + 1} V] [HasReverse V] {a b : V} : (a ⟶ b) → (b ⟶ a) :=
+  HasReverse.reverse'
+
+/-- A quiver `HasInvolutiveReverse` if reversing twice is the identity.`-/
+class HasInvolutiveReverse extends HasReverse V where
+  /-- `reverse` is involutive -/
+  inv' : ∀ {a b : V} (f : a ⟶ b), reverse (reverse f) = f
+
+
+@[simp]
+theorem reverse_reverse {V} [Quiver.{v + 1} V] [h : HasInvolutiveReverse V] {a b : V} (f : a ⟶ b) :
+    reverse (reverse f) = f := by apply h.inv'
+#align quiver.reverse_reverse Quiver.reverse_reverse
+
+@[simp]
+theorem reverse_eq_reverse_iff {V} [Quiver.{v + 1} V] [h : HasInvolutiveReverse V] {a b : V}
+    (f g : a ⟶ b) : reverse f = reverse g ↔ f = g := by
+  constructor
```

Comment:
```suggestion
  split
```

Reply:
I think I got an error on using `split` at first: it was not supported or something.

Reply:
Can you report that on Zulip?